From 46f0829ba9c914a2a804dfbdba30bb67e219df96 Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Tue, 27 Aug 2024 13:48:29 +0530
Subject: [PATCH] wifi: mac80211: Handle link sta removal during ML
 reconfiguration

While handling MLD reconfiguration work, sdata valid links will be
removed but sta_info valid links is not getting cleared. Because of
this, valid links in sta_info and sdata will be different which leads
to getting warn_on messages after a network restart or any disassociate
handling.

Fix this by removing the sta link after successfull MLD reconfiguration
removal.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 include/net/mac80211.h    |  2 ++
 net/mac80211/cfg.c        |  2 +-
 net/mac80211/driver-ops.h | 12 ++++++++++++
 net/mac80211/mlme.c       | 30 ++++++++++++++++++++++++++++--
 net/mac80211/sta_info.c   | 11 +++++++----
 net/mac80211/sta_info.h   |  2 +-
 6 files changed, 51 insertions(+), 8 deletions(-)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 5dbe6d3..b011d52 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -4843,6 +4843,8 @@ struct ieee80211_ops {
 	int (*link_reconfig_remove)(struct ieee80211_hw *hw,
 				    struct ieee80211_vif *vif,
 				    const struct cfg80211_link_reconfig_removal_params *params);
+	bool (*removed_link_is_primary)(struct ieee80211_sta *sta,
+					u16 removed_links);
 };
 
 /**
diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 2bb937d..2e67fa4 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -5738,7 +5738,7 @@ static int sta_add_link_station(struct ieee80211_local *local,
 
 	ret = sta_link_apply_parameters(local, sta, true, params);
 	if (ret) {
-		ieee80211_sta_free_link(sta, params->link_id);
+		ieee80211_sta_free_link(sta, params->link_id, false);
 		return ret;
 	}
 
diff --git a/net/mac80211/driver-ops.h b/net/mac80211/driver-ops.h
index 90efca0..8635055 100644
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -1585,4 +1585,16 @@ drv_link_reconfig_remove(struct ieee80211_local *local,
 
 	return ret;
 }
+
+static inline bool
+drv_check_removed_link_is_primary(struct ieee80211_local *local,
+				  struct sta_info *sta,
+				  u16 removed_links)
+{
+	if (local->ops->removed_link_is_primary)
+		return local->ops->removed_link_is_primary(&sta->sta,
+							   removed_links);
+
+	return false;
+}
 #endif /* __MAC80211_DRIVER_OPS */
diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 166c9fc..7c0cbe6 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -5768,6 +5768,8 @@ static void ieee80211_ml_reconf_work(struct wiphy *wiphy,
 	struct ieee80211_sub_if_data *sdata =
 		container_of(work, struct ieee80211_sub_if_data,
 			     u.mgd.ml_reconf_work.work);
+	struct sta_info *sta;
+	struct ieee80211_local *local = sdata->local;
 	u16 new_valid_links, new_active_links, new_dormant_links;
 	int ret;
 
@@ -5781,6 +5783,22 @@ static void ieee80211_ml_reconf_work(struct wiphy *wiphy,
 		   "MLO Reconfiguration: work: valid=0x%x, removed=0x%x\n",
 		   sdata->vif.valid_links, sdata->u.mgd.removed_links);
 
+	list_for_each_entry(sta, &local->sta_list, list) {
+		if (sdata != sta->sdata)
+			continue;
+
+		/* WAR: Until primary umac_migration is supported, disassoc the
+		 * entire MLD association if the link getting removed is
+		 * primary link in driver
+		 */
+		if (drv_check_removed_link_is_primary(local, sta,
+						      sdata->u.mgd.removed_links)) {
+			___ieee80211_disconnect(sdata);
+			sdata_unlock(sdata);
+			return;
+		}
+	}
+
 	new_valid_links = sdata->vif.valid_links & ~sdata->u.mgd.removed_links;
 	if (new_valid_links == sdata->vif.valid_links) {
 		sdata_unlock(sdata);
@@ -5818,9 +5836,17 @@ static void ieee80211_ml_reconf_work(struct wiphy *wiphy,
 		sdata_info(sdata, "Failed setting valid links\n");
 
 out:
-	if (!ret)
+	if (!ret) {
+		list_for_each_entry(sta, &local->sta_list, list) {
+			if (sdata != sta->sdata)
+				continue;
+
+			/* TODO: change remove links to link_id */
+			ieee80211_sta_free_link(sta, ffs(sdata->u.mgd.removed_links) - 1, true);
+		}
+
 		cfg80211_links_removed(sdata->dev, sdata->u.mgd.removed_links);
-	else
+	} else
 		___ieee80211_disconnect(sdata);
 
 	sdata->u.mgd.removed_links = 0;
diff --git a/net/mac80211/sta_info.c b/net/mac80211/sta_info.c
index 6488988..5b918a9 100644
--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -3045,11 +3045,11 @@ int ieee80211_sta_allocate_link(struct sta_info *sta, unsigned int link_id)
 	return 0;
 }
 
-void ieee80211_sta_free_link(struct sta_info *sta, unsigned int link_id)
+void ieee80211_sta_free_link(struct sta_info *sta, unsigned int link_id, bool unhash)
 {
 	lockdep_assert_held(&sta->sdata->local->sta_mtx);
 
-	sta_remove_link(sta, link_id, false);
+	sta_remove_link(sta, link_id, unhash);
 }
 
 int ieee80211_sta_activate_link(struct sta_info *sta, unsigned int link_id)
@@ -3110,8 +3110,11 @@ void ieee80211_sta_remove_link(struct sta_info *sta, unsigned int link_id)
 	sta->sta.valid_links &= ~BIT(link_id);
 
 	if (test_sta_flag(sta, WLAN_STA_INSERTED))
-		drv_change_sta_links(sdata->local, sdata, &sta->sta,
-				     old_links, sta->sta.valid_links);
+		if (drv_change_sta_links(sdata->local, sdata, &sta->sta,
+				     old_links, sta->sta.valid_links)) {
+			sta->sta.valid_links |= BIT(link_id);
+			return;
+		}
 
 	sta_remove_link(sta, link_id, true);
 }
diff --git a/net/mac80211/sta_info.h b/net/mac80211/sta_info.h
index 7d86edf..f131405 100644
--- a/net/mac80211/sta_info.h
+++ b/net/mac80211/sta_info.h
@@ -955,7 +955,7 @@ void ieee80211_sta_expire(struct ieee80211_sub_if_data *sdata,
 			  unsigned long exp_time);
 
 int ieee80211_sta_allocate_link(struct sta_info *sta, unsigned int link_id);
-void ieee80211_sta_free_link(struct sta_info *sta, unsigned int link_id);
+void ieee80211_sta_free_link(struct sta_info *sta, unsigned int link_id, bool unhash);
 int ieee80211_sta_activate_link(struct sta_info *sta, unsigned int link_id);
 void ieee80211_sta_remove_link(struct sta_info *sta, unsigned int link_id);
 
-- 
2.34.1

